pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +372,422 @@ pub mod pallet {
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapIdToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let pending_amount = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					SwapIdToForeignId::<T>::remove(swap_id);
+					T::Balance::zero()
+				}
+			};
+
+			ForeignInvestmentInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.increase_swap_id == Some(swap_id) {
+						T::Investment::update_investment(
+							&who,
+							investment_id,
+							T::Investment::investment(&who, investment_id)?
+								.ensure_add(last_swap.amount_in)?,
+						)?;
+
+						if pending_amount.is_zero() {
+							info.increase_swap_id = None;
+						}
+					} else if info.decrease_swap_id == Some(swap_id) {
+						T::DecreasedForeignInvestOrderHook::notify_status_change(
+							(who.clone(), investment_id),
+							ExecutedForeignDecreaseInvest {
+								amount_decreased: last_swap.amount_in,
+								foreign_currency: info.base.foreign_currency,
+								amount_remaining: pending_amount,
+							},
+						)?;
+
+						if pending_amount.is_zero() {
+							info.decrease_swap_id = None;
+						}
+					}
+				}
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			ForeignRedemptionInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.swap_id == Some(swap_id) {
+						info.swapped_amount.ensure_add_assign(last_swap.amount_in)?;
+
+						if info.base.collected.amount_collected == info.swapped_amount {
+							let tranche_tokens = info.pending_tranche_tokens_to_send()?;
+							info.tranche_tokens_sent.ensure_add_assign(tranche_tokens)?;
+
+							T::CollectedForeignRedemptionHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignCollect {
+									currency: info.base.foreign_currency,
+									amount_currency_payout: last_swap.amount_in,
+									amount_tranche_tokens_payout: tranche_tokens,
+									amount_remaining: info.remaining_tranche_tokens()?,
+								},
+							)?;
```

Comment:
I need to come back to this entire block tomorrow. Can't fully follow. 

For now, I would like to ask the question why we don't decrement `info.total_tranche_tokens`. AFAICS, this is missing such that `remaining_tranche_tokens` would never be zero right now.

Reply:
Yeah, this is the most tricky part. I'll try to re-thing if I could leave it more readable.

`total_tranche_tokens` is the amount the user expects to redeem. Incremented and decremented in `increase/decrease` methods.

When the redemption is collected [here](https://github.com/centrifuge/centrifuge-chain/pull/1698/files#diff-3d5d99980cfa0ce4827916a65ca67084aaee0f5b61eb6c07d6e086431604f93bR777) the tranche amount is aggregated in `info.collected.amount_payment`. So here in line `709`, we are actually doing: `amount_remaining = info.total_tranche_tokens - info.base.collected.amount_payment`

Reply:
No longer apply since [38347ff](https://github.com/centrifuge/centrifuge-chain/pull/1698/commits/38347ff110bd287b652bddea40e0eed147f4c121)